File: front/src/lib/entites/evenement.ts
/**
 * L'ÉVÉNEMENT de calendrier, déclaré une fois — les treize sections, ce qu'elles
 * portent, et **chaque divergence entre états avec son motif** (R4).
 *
 * Cinquième entité mise au cadre #430, et le lot **#432**.
 *
 * ## Pourquoi cet écran arrive maintenant, alors qu'il paraissait déjà conforme
 *
 * Le calendrier est le seul couple menu/entité du site à atteindre **100 % de
 * recouvrement entre création et édition** : un seul `form`, un seul `save()`,
 * `editId` pour seul discriminant, et `FormulaireEvenement` qui n'écrit aucun
 * champ commun à la main. Sa carte a rejoint la norme le 18/08/2026
 * (`EnteteCarte`, `ApercuCarte`, `HistoriqueEvenement`).
 *
 * 🔴 **Et pourtant le garde-fou y était vert par absence de sujet.**
 * `FormulaireEvenement` posait `avecPerimetre avecDescription avecPhotos
 * avecDocuments avecDiffusion` **en dur** — exactement ce que `lint:etats`
 * refuse — sans que rien n'échoue : le contrôle ignore tout fichier qui
 * n'importe aucune entité (`if (importe.length === 0) continue`). Un écran sans
 * déclaration n'est pas un écran conforme, c'est un écran **hors de portée du
 * contrôle**.
 *
 * C'est le cas zéro de `standards/04` §2 sous une autre forme : le contrôle
 * n'échouait pas, il n'avait simplement rien à regarder. Écrire ce fichier
 * l'allume.
 *
 * ## Qui consomme cette déclaration
 *
 *   • `FormulaireEvenement.svelte` — création ET édition (`sectionPresente`)
 *   • `CarteEvenement.svelte`      — l'affichage, via `FicheLecture`
 *   • `HistoriqueEvenement.svelte` → `EvolForm` — **pas encore** : voir la note
 *     en fin de fichier.
 */

import type { EntiteDeclaree } from './types';
import { DIFFUSION_NE_SE_LIT_PAS } from './types';

export const EVENEMENT: EntiteDeclaree = {
	id: 'evenement',
	libelle: 'Événement',
	libelleNouveau: 'Nouvel événement',
	libelleModifier: "Modifier l'événement",
	sections: [
		{
			id: 'titre',
			objet: 'Titre',
			requis: true,
			absente: {
				evolution: {
					motif: 'hérité',
					explication:
						"Une entrée de l'Historique se rattache à l'événement ; elle ne le renomme pas.",
				},
			},
		},
		{
			//  SEPT champs sous un titre de groupe — d'où « Détails » plutôt qu'une
			//  liste d'intitulés : une section à plusieurs champs garde son titre de
			//  groupe ET les libellés de ses champs, ce n'est pas une redite mais une
			//  hiérarchie (ux-patterns §9 septies).
			//
			//  ⚠️ La fréquence n'apparaît qu'avec un prestataire choisi, et sa valeur
			//  qu'avec une fréquence. C'est une divergence de CHAMP à l'intérieur
			//  d'une section, que R4 ne sait pas déclarer (#436) — elle est ici en
			//  commentaire, donc invisible au contrôle, et c'est dit.
			id: 'nature',
			objet: 'Type · Date de début · Heure · Fin · Lieu · Prestataire · Fréquence',
			titreEcran: 'Détails',
			absente: {
				evolution: {
					motif: 'hérité',
					explication:
						"Le type, les dates, le lieu et le prestataire appartiennent à l'événement. " +
						'Une entrée du fil raconte ce qui lui arrive, elle ne le reprogramme pas.',
				},
			},
			pliee: true,
		},
		{
			id: 'equipement',
			sansObjet:
				'Un événement du calendrier se rattache à un LIEU, pas à un équipement — il ' +
				"est dans « Détails ». L'équipement qualifie ce qu'on entretient, et c'est " +
				"l'affaire qui le porte.",
		},
		{
			//  🔴 LE KANBAN *EST* LE WORKFLOW (arbitré le 18/08/2026). Ses colonnes —
			//  AG · CS · Syndic · Prestataire · Terminé · Annulé — répondent exactement
			//  à la question de la section 3, « où en est cet objet ? ». La section
			//  s'appelait « Suivi Kanban », ce qui nommait l'écran où on le voit plutôt
			//  que la notion ; et avant cela elle était rangée dans la DIFFUSION, qui
			//  dit qui voit l'objet et non où il en est.
			//
			//  ⚠️ **Aucun second champ d'état n'existe** : `statut_kanban` et rien
			//  d'autre. Deux notions de suivi sur le même objet se contredisent au
			//  premier écart, et rien ne dirait laquelle fait foi.
			//
			//  Il est **présent en évolution**, et c'est ce qui a changé le 18/08 : le
			//  calendrier était le dernier écran du site à faire avancer un suivi en
			//  silence. Un changement de colonne est désormais une transition tracée,
			//  avec son avant et son après.
			//
			//  ⚠️ **Non requis**, à la différence de l'état d'un ticket : « — Pas de
			//  suivi » est une pastille comme les autres, active par défaut. Un
			//  événement peut légitimement n'avoir aucun suivi — une AG a une date,
			//  pas un dossier.
			id: 'suivi',
			objet: 'Kanban — AG · CS · Syndic · Prestataire · Terminé · Annulé (ou aucun suivi)',
			pliee: true,
		},
		{
			id: 'quand',
			sansObjet:
				"ANOMALIE, pas absence : un événement porte bien un début et une fin, mais ils vivent en section 2 (« Détails ») depuis toujours. C'est précisément parce qu'aucune section ne nommait la notion qu'il a fallu un TROISIÈME objet pour dire « ça se passe le X ». Le lot qui fait disparaître cette entité (#1092) résorbe la divergence en la supprimant — on ne la corrige donc pas ici.",
		},
		{
			//  Le prestataire d'un événement est dans « Détails » (`nature`) : il fait
			//  partie de ce qui définit l'intervention, au même titre que sa date.
			id: 'intervenant',
			sansObjet:
				"Le prestataire d'un événement vit dans sa section « Détails » (`nature`), " +
				"avec la date et le lieu : ensemble, ils DÉFINISSENT l'intervention.",
		},
		{
			id: 'perimetre',
			objet: 'PerimetrePicker — de quoi il s’agit',
			requis: true,
			absente: {
				evolution: {
					motif: 'hérité',
					explication: "Le périmètre est celui de l'événement ; une entrée ne le redéfinit pas.",
				},
			},
		},
		{
			//  ⚠️ **Non requise**, et c'est la seule des cinq entités déclarées dans ce
			//  cas. Un ticket sans description ne dit pas quel est le problème ; un
			//  événement dont on connaît le type, la date et le lieu se comprend sans
			//  une ligne de plus — « Coupure d'eau · bâtiment C · mardi » est complet.
			//  L'écran ne l'a jamais exigée : la déclaration le constate, elle ne le
			//  change pas.
			id: 'description',
			objet: 'RichEditor — de quoi il retourne',
			pliee: true,
		},
		{
			id: 'pieces_jointes',
			objet: 'FichiersUpload mode mixte — photos et documents',
			//  🔴 PLIÉE, comme toute section facultative (21/09/2026, demandé à
			//  l'écran). Elle était l'exception inverse — « dépliée parce que joindre
			//  une photo est le premier geste sur téléphone ». C'est l'ARBITRAGE qui a
			//  changé, pas la règle : un formulaire de treize sections toutes ouvertes
			//  n'est pas lisible au pouce.
			//  ⚠️ `valeurModifiee` la rouvre dès qu'un fichier y est joint : en
			//  édition, un objet qui porte des pièces jointes ne les cache pas.
			pliee: true,
		},
		{
			//  🔴 Obligatoire, donc DEPLIEE — et plus d'exception (22/09/2026).
			//
			//  Elle etait pliee, au motif que « en mon nom » est juste dans la
			//  quasi-totalite des cas. Signale a l'ecran le jour de sa livraison :
			//  une section marquee d'une asterisque qui arrive fermee demande de
			//  l'ouvrir pour voir qu'il n'y a rien a y faire, ce qui est exactement
			//  le contraire de ce que l'asterisque annonce.
			//
			//  ⚠️ C'etait la SEULE exception de pliage du cadre. La regle
			//  « obligatoire -> deplie, facultatif -> plie » n'en a plus aucune, et
			//  c'est ce que l'utilisateur a demande : « pour eviter toute exception ».
			id: 'au_nom_de',
			objet: 'Saisi pour — en mon nom · un résident inscrit · une personne extérieure',
			requis: true,
		},
		{
			id: 'destinataires',
			//  🔴 OBLIGATOIRE, donc DÉPLIÉE (22/09/2026, signé à l'écran deux fois).
			//
			//  L'astérisque était écrite en dur par `SectionDestinataires` : la
			//  déclaration ne savait donc pas que la section était obligatoire, et
			//  `lint:etats` — qui CALCULE le pliage à partir d'elle — ne voyait
			//  aucune contradiction à la déclarer pliée. L'écran affichait donc
			//  « DESTINATAIRES* » sur une ligne fermée, ce que la règle interdit.
			requis: true,
			sansObjet:
				"Un événement n'adresse personne nommément dans l'application : il est vu par qui " +
				'son périmètre concerne, et le Kanban filtre déjà ses colonnes selon le statut du ' +
				'lecteur. `envoyer_syndic` et `envoyer_cs` ne sont PAS des destinataires au sens ' +
				'de la section 5 — ce sont deux canaux, et ils vivent en Diffusion (section 9), ' +
				'comme sur les tickets.',
		},
		{
			id: 'mise_en_avant',
			objet: 'Épinglage · Urgence · Réservé au conseil · Confidentiel',
			pliee: true,
		},
		{
			//  ⚠️ Le TÉLÉVERSEMENT est **immédiat** sur cet écran, et ce n'est pas un
			//  détail d'implémentation : c'est ce qui permet à l'e-mail de notification
			//  de partir AVEC les photos. Rien ici ne doit le déplacer — et cela ne se
			//  voit pas à l'écran (#432).
			id: 'diffusion',
			objet:
				'CanauxNotification — WhatsApp, syndic, conseil syndical — et l’affichage au fil ' +
				'du tableau de bord (affichable, épinglé)',
			absente: {
				affichage: DIFFUSION_NE_SE_LIT_PAS,
			},
			pliee: true,
		},
	],
};

/**
 * ⚠️ **Ce que cette déclaration ne gouverne PAS encore, et pourquoi.**
 *
 * L'état `evolution` est déclaré ici, mais `EvolForm.svelte` ne le consomme
 * pas : ce composant sert **quatre écrans** — tickets, fiche de ticket,
 * actualités, espace CS — auxquels le calendrier s'est ajouté le 18/08/2026 via
 * `HistoriqueEvenement`. L'y brancher les changerait tous **avant** qu'on les
 * ait regardés, et c'est très exactement ce que R5 interdit : *l'enrichissement
 * se propose sur UN écran, se fait constater, puis se généralise*.
 *
 * L'écart concret qui subsiste est l'intitulé de la description, qui bascule
 * « Commentaire » / « Contenu » selon le mode — là où **R3** demande le même
 * libellé d'un formulaire à l'autre. Et l'ordre des sections y est tenu **par
 * recopie** de `ChampsCommuns`, donc par deux endroits qui doivent rester
 * d'accord sans qu'aucun contrôle ne le vérifie. Suivi en **#463**.
 *
 * ⚠️ **Une note fausse a vécu ici entre la v2.87.2 et la v2.87.4**, et elle
 * mérite d'être signalée plutôt qu'effacée : elle annonçait qu'`EvolForm` rendait
 * les pièces jointes « unifiées » (`separatePhotosAndDocs = false`), fusionnant
 * les sections 7 et 8. C'était vrai jusqu'au 18/08 au matin, et **faux au moment
 * où je l'ai écrite** — le mode avait disparu le jour même, avec son dernier
 * appelant. Je l'avais recopiée depuis `ticket.ts` sans la revérifier, et elle
 * citait de surcroît #433, fermé depuis. Un commentaire qui décrit un état révolu
 * coûte plus cher qu'un commentaire absent : il se cite avec assurance.
 */

File: api/app/models/evenement.py
"""Le CALENDRIER — un événement, et le fil de son suivi.

Créé le 18/08/2026, sur demande : *« pour Calendrier il doit y avoir un
historique et workflow »*. Le **workflow existait déjà** sous un autre nom — les
colonnes du Kanban (AG · CS · Syndic · Prestataire · Terminé · Annulé) répondent
exactement à la question de la section 3 du cadre #430, *« où en est cet
objet ? »*. Ce qui manquait, c'était la **trace** : une colonne changeait sans
que rien ne dise quand, par qui, ni pourquoi.

## Pourquoi un module à part

⚠️ Ce module ne portait que l'HISTORIQUE jusqu'au 19/08/2026 — `Evenement`
lui-même était resté dans `core.py`, qui a donc continué de grossir pendant
qu'un module portant son nom existait à côté. Les trois déclarations
(`TypeEvenement`, `StatutKanban`, `Evenement`) l’ont rejoint quand le garde-fou
a de nouveau refusé une ligne à `core.py` (#497).

⚠️ Aucun `Relationship` ici, et c'est ce qui rend le déplacement possible :
`Evenement` ne référence les autres tables que par clé étrangère nommée
(`batiment.id`, `utilisateur.id`, `prestataire.id`), ce qui n'impose aucun
cycle d'import. C'est la raison exacte pour laquelle `Ticket` et
`TicketEvolution`, eux, restent dans `core.py` (cf. `models/tickets.py`).

`core.py` fait 1 237 lignes et le garde-fou de modularité (rang 1,
`standards/02` §6) refuse qu'un fichier au-delà de 500 grossisse pour une
nouvelle fonctionnalité. Même raison que `models/tickets.py` (17/08) et
`models/copropriete.py` (13/08).

⚠️ Ce module doit être **importé** par `models/__init__.py` : un modèle que
personne n'a importé n'existe pas pour `metadata.create_all`, et la table
manquerait sans le moindre message.

## La forme est celle des deux autres fils, et c'est délibéré

`TicketEvolution` et `PublicationEvolution` portent les mêmes colonnes. Les trois
alimentent la **même** rubrique d'affichage (`RubriqueHistorique.svelte`), qui ne
connaît aucune entité — elle attend `type`, `contenu`, `ancien_statut`,
`nouveau_statut`, `auteur_nom`, `cree_le`, `fichiers_urls`. Diverger ici aurait
obligé à lui ajouter une variante, et *une variante ajoutée pour accueillir un
écart existant ne factorise pas : elle entérine*.
"""
from datetime import datetime
from enum import Enum
from typing import Optional

from sqlmodel import Field, SQLModel


class TypeEvenement(str, Enum):
    travaux = "travaux"
    coupure = "coupure"
    ag = "ag"
    maintenance = "maintenance"
    maintenance_recurrente = "maintenance_recurrente"
    autre = "autre"


class StatutKanban(str, Enum):
    ag = "ag"
    cs = "cs"
    syndic = "syndic"
    fournisseur = "fournisseur"
    termine = "termine"
    annule = "annule"


class Evenement(SQLModel, table=True):
    __tablename__ = "evenement"
    id: Optional[int] = Field(default=None, primary_key=True)
    titre: str
    description: Optional[str] = None
    type: TypeEvenement = TypeEvenement.autre
    lieu: Optional[str] = None
    debut: datetime
    fin: Optional[datetime] = None
    perimetre: str = "résidence"  # résidence | bâtiment
    batiment_id: Optional[int] = Field(default=None, foreign_key="batiment.id")
    auteur_id: int = Field(foreign_key="utilisateur.id")
    cree_le: datetime = Field(default_factory=datetime.utcnow)
    mis_a_jour_le: Optional[datetime] = None
    archivee: bool = False
    statut_kanban: Optional[str] = Field(default=None)  # ag|cs|syndic|fournisseur|termine|annule
    prestataire_id: Optional[int] = Field(default=None, foreign_key="prestataire.id")
    #  🔴 La SOURCE de la visite, quand elle vient du pré-remplissage du kanban
    #  (#605, point 2). Le rapprochement se faisait sur le titre littéral : le
    #  renommer — ou renommer le prestataire — recréait tout l'exercice en double.
    #
    #  ⚠️ Nullable, et ça le reste : les visites créées AVANT le 01/09/2026 n'en
    #  ont pas, et un événement de maintenance saisi à la main n'a pas de contrat.
    #  ⚠️ PAS de `foreign_key=` : SQLite ne sait pas ajouter une contrainte à une
    #  table existante, et la migration 0165 a crashé en production pour l'avoir
    #  tenté. Une base neuve et une base migrée doivent porter le même schéma —
    #  le déclarer ici et pas là ferait diverger les deux en silence.
    contrat_id: Optional[int] = Field(default=None)
    frequence_type: Optional[str] = Field(default=None)   # "semaines", "mois", "fois_par_an"
    frequence_valeur: Optional[int] = Field(default=None)
    affichable: bool = Field(default=False)  # visible dans le dashboard (évènements récents)
    # Même convention que Ticket.photos_urls (tableau JSON d'URLs internes) : le
    # modèle portait déjà trois noms pour la même notion, ne pas en créer un 4ᵉ.
    photos_urls: Optional[str] = None  # JSON array of photo URLs
    # Pièces jointes non-images, même convention que Ticket.fichiers_urls.
    fichiers_urls: str = "[]"  # JSON array d'URLs de fichiers joints
    epingle: bool = False  # même notion que Publication.epingle
    partager_whatsapp: bool = False
    envoyer_syndic: bool = False
    envoyer_cs: bool = False


class EvenementEvolution(SQLModel, table=True):
    __tablename__ = "evenement_evolution"

    id: Optional[int] = Field(default=None, primary_key=True)
    evenement_id: int = Field(foreign_key="evenement.id")
    #  type : commentaire | etat — et deux seulement, comme pour les tickets et
    #  les publications. Une CORRECTION s'inscrit en `commentaire` préfixé
    #  « Correction : » ; seule une transition volontaire du Kanban est un `etat`.
    type: str
    contenu: Optional[str] = None
    #  Les deux colonnes du Kanban, avant et après. Renseignées pour un `etat`,
    #  nulles pour tout le reste — c'est leur absence qui distingue une correction
    #  d'une étape du suivi, et c'est ce que le fil lit pour ne pas dessiner un
    #  jalon là où il n'y en a pas.
    ancien_statut: Optional[str] = None
    nouveau_statut: Optional[str] = None
    auteur_id: int = Field(foreign_key="utilisateur.id")
    cree_le: datetime = Field(default_factory=datetime.utcnow)
    #  Tableau JSON d'URLs internes — même convention que `Ticket.photos_urls`.
    fichiers_urls: str = "[]"
